languages/rust/src/polar.rs

```diff
@@ -0,0 +1,127 @@
+/// Communicate with the Polar virtual machine: load rules, make queries, etc/
+use std::fs::File;
+use std::io::Read;
+use std::rc::Rc;
+use std::sync::{Arc, Mutex};
+
+use crate::host::Host;
+
+#[derive(Clone)]
+pub struct Polar {
+    inner: Rc<crate::PolarCore>,
+    host: Arc<Mutex<Host>>,
```

Comment:
I'm concerned all the locking on host will mean queries across threads (since they share a host) will have significant contention.  Can we either move locking of data structures internal to Host so that callers do not need to manage calling `lock()` and we can have more granular locking only on portions of operations that need it?

Alternatively, should we consider having a single host class per query?

Is this thread-safety issue something we may have missed when implementing this in other languages?

Reply:
Same comment here: https://github.com/osohq/oso/pull/391/files/33fe4fb3bedac9a9ff6b302ddba03a23e94e2cd9#r482588557

Let's revisit. I'm sure there are plenty of improvements on this, but this was the "good enough" version without worrying about borrows + references.